src/TimeSteppers/generic_time_stepping.jl

```diff
@@ -84,16 +86,18 @@ Update the horizontal velocities u and v via
 
 Note that the vertical velocity is not explicitly time stepped.
 """
-function _fractional_step_velocities!(U, grid, Δt, pNHS)
-    @loop_xyz i j k grid begin
-        @inbounds U.u[i, j, k] -= ∂xᶠᵃᵃ(i, j, k, grid, pNHS) * Δt
-        @inbounds U.v[i, j, k] -= ∂yᵃᶠᵃ(i, j, k, grid, pNHS) * Δt
-    end
-    return nothing
+@kernel function _fractional_step_velocities!(U, grid, Δt, pNHS)
+    i, j, k = @index(Global, NTuple)
+
+    @inbounds U.u[i, j, k] -= ∂xᶠᵃᵃ(i, j, k, grid, pNHS) * Δt
+    @inbounds U.v[i, j, k] -= ∂yᵃᶠᵃ(i, j, k, grid, pNHS) * Δt
 end
 
 "Update the solution variables (velocities and tracers)."
 function fractional_step_velocities!(U, C, arch, grid, Δt, pNHS)
-    @launch device(arch) config=launch_config(grid, :xyz) _fractional_step_velocities!(U, grid, Δt, pNHS)
+    workgroup, worksize = work_layout(grid, :xyz)
+    kernel! = _fractional_step_velocities!(device(arch), workgroup, worksize)
+    event = kernel!(U, grid, Δt, pNHS)
+    wait(event)
```

Comment:
You will have to apply the pattern consistently otherwise you open yourself up for races from above.
```suggestion
    wait(device(arch), event)
```